neo/SmartContract/InteropService.cs

```diff
@@ -143,8 +143,30 @@ private static bool Runtime_GetTrigger(ApplicationEngine engine)
 
         internal static bool CheckWitness(ApplicationEngine engine, UInt160 hash)
         {
-            var _hashes_for_verifying = engine.ScriptContainer.GetScriptHashesForVerifying(engine.Snapshot);
-            return _hashes_for_verifying.Contains(hash);
+            if (engine.ScriptContainer?.Witnesses == null) return false;
+
+            foreach (var witness in engine.ScriptContainer.Witnesses)
+            {
+                if (witness.ScriptHash != hash) continue;
+
+                switch (witness.Scope.Type)
+                {
+                    default: return false;
+
+                    case WitnessScopeType.Global: return true;
+                    case WitnessScopeType.CustomScriptHash: return engine.CurrentScriptHash == new UInt160(witness.Scope.ScopeData);
+                    case WitnessScopeType.RootAccess: return engine.CurrentScriptHash == engine.EntryScriptHash;
+                    case WitnessScopeType.ExecutingGroupPubKey:
```

Comment:
This is really beautiful.
I would ask if `CustomScriptHash` should be a vector, but, then, I understood that `ExecutingGroupPubKey` will kind of be this vector, right? It will contain a list of all trusted for that group, correct?

Reply:
Yes, this will come from manifest. The idea is that, if someone an operation that depends on Group permission, we could allow that. Usually, I think that natural protection of groups would do this, but it's better `intent` to inform this explicitly, than just putting Global and relying on external security.